# Comparação de Métodos Numéricos

Este notebook tem como objetivo determinar o valor de entrada `n` em que os algoritmos A e B apresentam o mesmo tempo de execução.

As funções que representam os tempos de execução são:

$$
T_A(n) = 0.01n\log_2(n)
$$

$$
T_B(n) = 0.08n + 20
$$

Para determinar o ponto em que os tempos são iguais, buscamos a raiz da função:

$$
f(n) = T_A(n) - T_B(n)
$$

ou seja:

$$
f(n) = 0.01n\log_2(n) - 0.08n - 20
$$

In [42]:
import math
import pandas as pd
import matplotlib.pyplot as plt

In [43]:
def TA(n):
    return 0.01 * n * math.log2(n)

def TB(n):
    return 0.08 * n + 20

def f(n):
    return TA(n) - TB(n)

In [44]:
print("n = 10:", f(10))
print("n = 100:", f(100))
print("n = 1000:", f(1000))

n = 10: -20.467807190511266
n = 100: -21.356143810225277
n = 1000: -0.34215715337913366


In [45]:
for n in range(1, 10001):
    if f(n) * f(n + 1) < 0:
        print(f"Mudança de sinal entre n = {n} e n = {n + 1}")

Mudança de sinal entre n = 1010 e n = 1011


In [ ]:
import sys
from pathlib import Path

sys.path.append(str(Path.cwd().parent / "src"))

from metodos.posicao_falsa import posicao_falsa

In [ ]:
raiz_pf, iteracoes_pf = posicao_falsa(
    f,
    1010,
    1011,
    1e-6,
    1e-6,
    100
)

print("Método da Posição Falsa")
print("n encontrado:", raiz_pf)
print("Iterações:", iteracoes_pf)
print("Resíduo:", abs(f(raiz_pf)))

Método da Posição Falsa
n encontrado: 1010.0172413456174
Iterações: 1
Resíduo: 1.2097495982743567e-07


### Resultado - Método da Posição Falsa

O Método da Posição Falsa encontrou a raiz aproximada:

$$
n \approx 1010.017241
$$

O método convergiu em **1 iteração**, apresentando resíduo de aproximadamente:

$$
1.21 \times 10^{-7}
$$

Como o resíduo obtido é inferior à tolerância estabelecida de $10^{-6}$, o resultado satisfaz o critério de precisão utilizado.

Esse valor representa aproximadamente o tamanho de entrada em que os algoritmos A e B apresentam o mesmo tempo de execução.